File: app/src/main/java/com/example/liars_dice/api/GameAPI.java
package com.example.liars_dice.api;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.URISyntaxException;

import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

public class GameAPI {
    private static GameAPI INSTANCE = null;

    public enum Event {
        DISCONNECT("disconnect"),
        UPDATE("game:state:update"),
        LOST("game:lost"),
        WON("game:won");


        private final String text;

        Event(String text) {
            this.text = text;
        }

        @Override
        public String toString() {
            return this.text;
        }
    }

    private Socket socket = null;

    public static GameAPI getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new GameAPI();
        }
        return INSTANCE;
    }

    public void connect(String id) {
        System.out.println("Connecting with id " + id);
        try {
            socket = IO.socket("http://10.0.2.2:3000/games/"+id);
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        socket.connect();
    }

    public void on(Event event, Emitter.Listener listener) {
        this.socket.on(event.toString(), listener);
    }

    public void off(Event event, Emitter.Listener listener) {
        this.socket.off(event.toString(), listener);
    }

    public String getID() {
        return this.socket.id();
    }

    public void bet(Integer amount, Integer value) {
        System.out.println("API doing bet stuff");
        JSONObject data = new JSONObject();
        try {
            data.put("amount", amount);
            data.put("value", value);
            System.out.println("API set bet data");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        System.out.println("API emiting bet");
        this.socket.emit("game:action:bet",data);
    }

    public void call() {
        System.out.println("API emitting call");
        this.socket.emit("game:action:call");
    }
}

File: server/src/apis/GameAPI.ts
import { Namespace } from "socket.io";
import { GameService, GameState } from "../services/GameService";
import { Bet } from "../models/game/bet";
import { Server } from "../server";
import { Result } from "../utils";



export class GameAPI {
    static EVENT = {
        BET: "game:action:bet",
        CALL: "game:action:call",
        DISCONNECT: "disconnect"
    }
    private server: Server;
    private id: string;
    private nsp: Namespace;
    private controller: GameService;

    constructor(id: string, controller: GameService) {
        this.server = Server.getInstance();
        this.id = id;
        this.nsp = this.server.createNameSpace(`games/${this.id}`);
        this.controller = controller;
    }

    listen() {
        this.nsp.on("connect", socket => {
            console.log(socket.id, "joined the game ", this.id);
            this.controller.addPlayer(socket.id);
            socket.on("disconnect", ()=> {
                // TODO: Disconnect stuff
            })

            socket.on("game:action:bet", args => {
                console.log(socket.id, " wants to make a bet")
                const {amount,value} = args;
                if (amount === undefined || value === undefined) {
                    return;
                }
                const {data,error} = this.controller.makeBet(socket.id, new Bet(value, amount));
                if (error) {
                    return;
                }
                if (data === undefined) {
                    return;
                }
                const {data: state} = this.controller.makeBet(socket.id, new Bet(value,amount));
                if (state !== undefined) {
                    this.updateState(state);

                }
            }) 

            socket.on("game:action:call" , () => {
                console.log(socket.id, " wants to call")
                const {data, error} = this.controller.call(socket.id);
                if (error !== undefined) {
                    return;
                }
                if (data === undefined) {
                    return;
                }
            })
        })
    }

    playerWins(id: string) {
        console.log("gameAPI running playerWins")
        const socket = this.nsp.sockets.get(id);
        if (socket !== undefined) {
            socket.emit("game:won");
        }
    }

    playerLost(id: string) {
        console.log("gameAPI running playerLost")
        const socket = this.nsp.sockets.get(id);
        if (socket !== undefined) {
            socket.emit("game:lost");
        }
    }

    updateState(data: GameState) {
        console.log("sending game state ", data)
        const result = {data}
        this.nsp.emit("game:state:update", result);
    }

    deleteNameSpace() {
        this.server.deleteNameSpace(this.nsp);
    }
}